cmake: Install internal binaries to <prefix>/libexec/

ryanofsky opened the pull request:
This change moves binaries that are not typically invoked directly by users from the `bin/` directory to the `libexec/` directory in CMake installs and binary releases. The goal of the PR is to introduce a distinction between internal and external binaries so starting with #31802, we can use IPC to implement features in new binaries without adding those binaries to the CLI. The change also helps reduce clutter in `bin/`, making it easier for users to identify useful tools to run. Summary of changes:

- For **source builds** (i.e. developer builds) — There are no changes.
- For **source installs** (i.e. `cmake --install` result) — `test_bitcoin`, `test_bitcoin-qt`, and `bench_bitcoin` are installed in `${CMAKE_PREFIX_PATH}/libexec` instead of `${CMAKE_PREFIX_PATH}/bin`, so they are no longer on the system `PATH`. However, they can still be invoked from the `libexec/` directory, or from the CLI as `bitcoin test`, `bitcoin test-gui`, and `bitcoin bench`, respectively.
- For **binary releases** — Since `test_bitcoin` is the only test binary enabled in releases, the only change is moving `test_bitcoin` from `bin/` to `libexec/`.

<details><summary>Details</summary>
<p>

 The table below shows the install location of each binary after this change, and the availability of each binary.

| Binary               | Location     | Availability         | Change                        |
|----------------------|--------------|----------------------|-------------------------------|
| `bitcoin`            | `bin/`       | 📦 Binary release (since #31375) | Unchanged                     |
| `bitcoin-cli`        | `bin/`       | 📦 Binary release     | Unchanged                     |
| `bitcoind`           | `bin/`       | 📦 Binary release     | Unchanged                     |
| `bitcoin-qt`         | `bin/`       | 📦 Binary release     | Unchanged                     |
| `bitcoin-tx`         | `bin/`       | 📦 Binary release     | Unchanged                     |
| `bitcoin-util`       | `bin/`       | 📦 Binary release     | Unchanged                     |
| `bitcoin-wallet`     | `bin/`       | 📦 Binary release     | Unchanged                     |
| `bench_bitcoin`      | `libexec/`   | 🛠 Source build only  | Moved from `bin/`             |
| `bitcoin-chainstate` | `libexec/`   | 🛠 Source build only  | Newly installed (was built)   |
| `bitcoin-gui`        | `libexec/`   | 🛠 Source build only (until #31802) | Moved from `bin/`             |
| `bitcoin-node`       | `libexec/`   | 🛠 Source build only (until #31802) | Moved from `bin/`             |
| `test_bitcoin`       | `libexec/`    | 📦 Binary release     | Moved from `bin/`             |
| `test_bitcoin-qt`    | `libexec/`   | 🛠 Source build only  | Moved from `bin/`             |

</p>
</details>

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
cmake: Move internal binaries from bin/ to libexec/

This change moves binaries that are not typically invoked directly by users
from the `bin/` directory to the `libexec/` directory in CMake installs and
binary releases. The goal is to simplify the contents of `bin/` for end users
while still making all binaries available when needed. After this change, the
binaries remaining in `bin/` are:

- bitcoin
- bitcoin-cli
- bitcoind
- bitcoin-qt
- bitcoin-tx
- bitcoin-util
- bitcoin-wallet

And the binaries that are moved to `libexec/` are:

- bench_bitcoin
- bitcoin-chainstate(*)
- bitcoin-gui(***)
- bitcoin-node(***)
- test_bitcoin(**)
- test_bitcoin-qt

(*) bitcoin-chainstate was previously missing an install rule and was actually
not installed even when it was enabled.

(**) test_bitcoin is the only libexec/ binary that is currently included in
bitcoin binary releases. The others are only installed when building from
source with relevant cmake options enabled.

(***) bitcoin-node and bitcoin-gui are not currently built by default or
included in binary releases but both of these changes are planned and
implemented in #31802

luke-jr commented:
`libexec` is usually used for binaries run by programs rather than people. I don't think most of the ones you're moving fit that, and should remain in `bin`.

Sjors commented:
If there are no binaries for which it currently makes sense to put them in `libexec`, since they're all called by people, then it seems this PR has be based on #31375.

ryanofsky commented:
The idea that an executable placed in `libexec/` should be called by other executables is a good rule of thumb but not a requirement of some kind. (And if it were a requirement, it would be satisfied by #31375.)

The technical difference between binaries installed to `bin/` and binaries installed to `libexec/` is that binaries in `bin/` get exposed on the `PATH` while binaries in `libexec/` don't.

There is nothing wrong with binaries in `libexec/` being documented and called used by users and packagers and other tools and scripts. Not every executable that a user could conceivably call needs to be present in `PATH`.

Packages like [postfix](https://www.postfix.org/postfix-manuals.html) have manpages for internal executables installed in `libexec/` ([anvil](https://www.postfix.org/anvil.8.html), [bounce](https://www.postfix.org/bounce.8.html), [smtpd](https://www.postfix.org/smtpd.8.html), etc). Git subcommands can be called manually from `libexec` and have manpages of their own, too. I would say postfix and git are pretty much archetypes of packages designed around unix principles and there is nothing novel about installing low-level executables in `libexec/`.

With this PR, two sets of binaries are installed to `libexec/` if they have been enabled in cmake:

1. multiprocess binaries `bitcoin-node` and `bitcoin-gui`
2. testing binaries `bench_bitcoin` `test_bitcoin` `test_bitcoin-qt`.

I can't think of a reason why multiprocess binaries should be installed on PATH, but if someone wants to make that case I'm ready to hear it. As an alternative to this PR, they could be installed in `bin/` with different names like `bitoin-node-multiprocess` or to a completely new folder. 

For the test and bench binaries I just think it's confusing and not useful to install these on the PATH. I think this is true today and will be especially true after #31375.

ryanofsky commented:
Updated 2a952710661a1990ed1c24b0ebd16de8ac0df87c -> 47a872236e070814ad74922d3f8a653e1c6af968 ([`pr/libexec.1`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.1) -> [`pr/libexec.2`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.1..pr/libexec.2)) to fix CI error

ryanofsky commented:
Rebased 99f2f335b6a8c2b5fe518a0c0056eb2a42c0496b -> 9194e461518e6b843c1a707ba1b10ab7a000e096 ([`pr/mine.15`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.15) -> [`pr/mine.16`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.15-rebase..pr/mine.16)) due to conflict with #31834. Also added status line for the executable as suggested

ryanofsky commented:
Rebased d2ceb2e0735a2c8343f8316b55fac55323aba62c -> 45bfd97ec7c9991f41673d79b01277bfd940e64e ([`pr/libexec.3`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.3) -> [`pr/libexec.4`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.3-rebase..pr/libexec.4)) due to conflict with #31844

TheCharlatan commented:
Concept ACK

jonatack commented:
Concept ACK

Would need a release note?

ryanofsky committed:
doc: Add release notes describing libexec/ binaries

ryanofsky committed:
doc: Add description of installed files to files.md

ryanofsky commented:
> Would need a release note?

Thanks! Added release note and files.md documention in new commits.

---

Added 2 commits 45bfd97ec7c9991f41673d79b01277bfd940e64e -> bda7626e987ce3d7287f72e6989dedde2c4ab666 ([`pr/libexec.4`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.4) -> [`pr/libexec.5`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.4...pr/libexec.5)) with release notes and documentation

TheCharlatan commented:
I think some installation packagers run `test_bitcoin` as part of their installation recipes, or at least I found one such example: https://github.com/epiccurious/bitcoin-core-node-builder/pull/164/files. @hebasto also polled usage of  `test_bitcoin` on social media: https://x.com/hhebasto/status/1729879307768738070. Clearly there aren't many people doing this, but are we ok with breaking the workflows of the few people that are? You mention that:

> For the test and bench binaries I just think it's confusing and not useful to install these on the PATH. I think this is true today and will be especially true after https://github.com/bitcoin/bitcoin/pull/31375.

What is confusing about keeping them where there are for now? Maybe there is some utility in testing your platform before actually running bitcoin on it.

ryanofsky commented:
Rebased bda7626e987ce3d7287f72e6989dedde2c4ab666 -> d253fa9ebe4305118417a02aa72cc06810662ac6 ([`pr/libexec.5`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.5) -> [`pr/libexec.6`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.5-rebase..pr/libexec.6)) due to conflict with #28710

Sjors commented:
Just #28710, no conflict with #31375? Will re-review shortly.

ismaelsadeeq reviewed (commented):
Concept ACK

hodlinator commented on src/CMakeLists.txt:
```diff
@@ -469,6 +469,7 @@ if(BUILD_UTIL_CHAINSTATE)
       core_interface
       bitcoinkernel
   )
+  install_binary_component(bitcoin-chainstate INTERNAL)
```
Do we want to *introduce* installation of this binary? Should be called out in PR description and/or commit message if kept.

Sjors replied:
I think it's fine, but indeed should be in the description.

TheCharlatan replied:
If so, it should also get a note that its command line interface is not expected to be stable.

fanquake replied:
> I think it's fine,

Not sure I agree that we should start installing this. It's an experimental, demo binary, not built by default, that nobody is using; other than a handful of developers working on the kernel (also experimental). There's every likelyhood it could just be removed tommorow.

hodlinator reviewed (commented):
Concept ACK d253fa9ebe4305118417a02aa72cc06810662ac6

### Concept

Would be nice (but not critical) to get this done in the same release where we introduce the now merged wrapper, bitcoin(.exe).

### Tested

```
₿ cmake --install build --prefix $PWD/install |sort
-- Install configuration: "RelWithDebInfo"
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoin
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoin-cli
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoind
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoin-qt
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoin-tx
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoin-util
-- Installing: /home/hodlinator/bc/b2/install/bin/bitcoin-wallet
-- Installing: /home/hodlinator/bc/b2/install/lib64/libbitcoinkernel.so
-- Installing: /home/hodlinator/bc/b2/install/lib64/pkgconfig/libbitcoinkernel.pc
-- Installing: /home/hodlinator/bc/b2/install/libexec/bench_bitcoin
-- Installing: /home/hodlinator/bc/b2/install/libexec/bitcoin-chainstate
-- Installing: /home/hodlinator/bc/b2/install/libexec/bitcoin-gui
-- Installing: /home/hodlinator/bc/b2/install/libexec/bitcoin-node
-- Installing: /home/hodlinator/bc/b2/install/libexec/test_bitcoin
-- Installing: /home/hodlinator/bc/b2/install/libexec/test_bitcoin-qt
-- Installing: /home/hodlinator/bc/b2/install/share/man/man1/bitcoin-cli.1
-- Installing: /home/hodlinator/bc/b2/install/share/man/man1/bitcoind.1
-- Installing: /home/hodlinator/bc/b2/install/share/man/man1/bitcoin-qt.1
-- Installing: /home/hodlinator/bc/b2/install/share/man/man1/bitcoin-tx.1
-- Installing: /home/hodlinator/bc/b2/install/share/man/man1/bitcoin-util.1
-- Installing: /home/hodlinator/bc/b2/install/share/man/man1/bitcoin-wallet.1
-- Set non-toolchain portion of runtime path of "/home/hodlinator/bc/b2/install/libexec/bitcoin-chainstate" to ""
```

```
./install/bin/bitcoin node (bin)
./install/bin/bitcoin test (libexec)
```

ismaelsadeeq reviewed (commented):
Tested ACK d253fa9ebe4305118417a02aa72cc06810662ac6

By building with bench in a release build
```terminal
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-cli
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-tx
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-util
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-wallet
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoind
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/libexec/bench_bitcoin
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/libexec/test_bitcoin
```

Sjors commented:
Now that the wrapper binary landed in #31375 I think it's fine to ask users to use `bitcoin test` instead of `test_bitcoin`.

I agree `test_bitcoin` (and maybe `bench_bitcoin`) are useful enough to have them in the release.

But I doubt there are many _automated_ deployments out there that obtain the release binaries and run the tests. That's the only thing this PR would break.

janb84 reviewed (commented):
Concept ACK d253fa9ebe4305118417a02aa72cc06810662ac6

This PR changes the target of some executables on install (different folder). This achieves that not all the executables are added to the PATH (which I agree with) 

Result of install ✅
<details>

```shell
cmake --install build --prefix $PWD/install
-- Install configuration: "RelWithDebInfo"
-- Installing: /Users/jan/Projects/bitcoin/install/bin/bitcoin-wallet
-- Installing: /Users/jan/Projects/bitcoin/install/share/man/man1/bitcoin-wallet.1
-- Installing: /Users/jan/Projects/bitcoin/install/bin/bitcoin
-- Installing: /Users/jan/Projects/bitcoin/install/bin/bitcoind
-- Installing: /Users/jan/Projects/bitcoin/install/share/man/man1/bitcoind.1
-- Installing: /Users/jan/Projects/bitcoin/install/libexec/bitcoin-node
-- Installing: /Users/jan/Projects/bitcoin/install/bin/bitcoin-cli
-- Installing: /Users/jan/Projects/bitcoin/install/share/man/man1/bitcoin-cli.1
-- Installing: /Users/jan/Projects/bitcoin/install/bin/bitcoin-tx
-- Installing: /Users/jan/Projects/bitcoin/install/share/man/man1/bitcoin-tx.1
-- Installing: /Users/jan/Projects/bitcoin/install/bin/bitcoin-util
-- Installing: /Users/jan/Projects/bitcoin/install/share/man/man1/bitcoin-util.1
-- Installing: /Users/jan/Projects/bitcoin/install/libexec/test_bitcoin
```
</details>

NIT; would change the wording of this sentence in the PR description, imho it's overly harsh for no reason: 

"Removing these esoteric and not generally useful binaries from bin/ and from the system PATH should make it easier for typical users to find and identify the binaries which are actually useful."

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated d253fa9ebe4305118417a02aa72cc06810662ac6 -> 705791cd436f237fe9bbac2cf52d63ab4b2a41c7 ([`pr/libexec.6`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.6) -> [`pr/libexec.7`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.6..pr/libexec.7)) just fixing typos pointed out in reviews.

---

re: TheCharlatan  https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-2807539249

I agree it is nice to distribute tests and I imagine in future we may want to distribute other diagnostic tools like this as well (especially if we start using the kernel as originally anticipated and writing specialized utilities with it, maybe even in other languages like rust.)

I just think it is good to make a distinction between common, everyday commands and more specialized ones, and only install the common commands we think will be used on a regular basis in `bin/` and in the system PATH, so they are discoverable and not mixed in with a list of more obscure functions. As mentioned [previously](https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-2619701583) I think git and postfix are good examples of unix packages that have a lot of internal binaries, and document them well, and allow them to be called freely, but are selective about what gets installed in `bin/` on the system `PATH`.

As far as breaking workflows, this change does include release notes, and I feel like it should be pretty easy for anyone with a custom verification script to update it to call `bitcoin test` or use `libexec/` instead of `bin/`. I don't think it should be a big deal to move a binary that wasn't previously documented to a new location, especially if it's part of a larger effort to organize and document all the binaries better and provide a more accessible interface in the form of the `bitcoin` command.

I could tweak the PR to leave `test_bitcoin` in `bin/` instead of `libexec/` or leave everything except `bitcoin-node` and `bitcoin-gui` in `bin/`. It just seemed nicer draw a distinction between `bin/` and `libexec/` based on what a good command line interface would look like, not based on idiosyncracies of previous releases.

---

re: Sjors https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-2922787970

Seems like a good suggestion to improve `bitcoin` help output. I will try implementing that

---

re: fanquake  https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-3027233699

Yes, sorry, removed that comment about not being generally useful. I think they are useful and think we should add more utilities like this, I think they just don't really make sense being top level commands on the system PATH instead of internal binaries and `bitcoin` subcommands.

---

re: janb84 https://github.com/bitcoin/bitcoin/pull/31679#pullrequestreview-2978891971

Thank you! Followed up on your suggestion and rewrote the PR description.

hodlinator reviewed (approved):
ACK 705791cd436f237fe9bbac2cf52d63ab4b2a41c7

Fixed typos and improved PR description since my first (concept & test) review (https://github.com/bitcoin/bitcoin/pull/31679#pullrequestreview-2892807188).

janb84 reviewed (commented):
re ACK 705791cd436f237fe9bbac2cf52d63ab4b2a41c7

Changes sinds last review: 

- Fixed typos 
- Changed PR description  (thanks! )

Sjors commented on doc/release-notes-31679.md:
```diff
@@ -0,0 +1,11 @@
+`bitcoin-node`, and `bitcoin-wallet` binaries are now installed in
+`$PREFIX/libexec` rather than `$PREFIX/bin`. If you are using a binary release
+or building from source with default build options, most of these programs are
+not built by default anyway so these changes may not be noticeable.
```
04e51ba1110e13598e0b11a35ce5abf4a1789f53: now that the wrapper binary is available, the release note could enumerate how to call these now: `bitcoin test`, etc.

Sjors reviewed (approved):
re-ACK 705791cd436f237fe9bbac2cf52d63ab4b2a41c7

Left a suggestion for the release note.

ismaelsadeeq commented:
reACK 705791cd and tested it again by building with release config.
```terminal
-- Install configuration: "Release"
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-wallet
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/share/man/man1/bitcoin-wallet.1
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoind
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/share/man/man1/bitcoind.1
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-cli
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/share/man/man1/bitcoin-cli.1
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-tx
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/share/man/man1/bitcoin-tx.1
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/bin/bitcoin-util
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/share/man/man1/bitcoin-util.1
-- Installing: /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin/install/libexec/test_bitcoin
```

I agree with the reply in https://github.com/bitcoin/bitcoin/pull/31679#pullrequestreview-2979400065

Also nice PR description update that makes it explicit that`test_bitcoin` is still available in release build but in `libexec` directory I also verify it in my build.

ryanofsky committed:
bitcoin wrapper: improve help output

Try to make extra commands more obvious based on a suggestion from Sjors:
https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-2922787970i

When `bitcoin` is invoked with no arguments, still show short help output, but
now explicitly state that more commands are available and `bitcoin help` will
list them.

Also:

- Get rid of -a/--all option. Just show all commands when `bitcoin help` or
  `bitcoin --help` is used. It maybe a helpful to add an option like this if
  more commands are added in the future, but right now there are not very many.

- Just show name of executable, not full path of executable in help output. This
  can be a little easier to read if the path is long.

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated 705791cd436f237fe9bbac2cf52d63ab4b2a41c7 -> 05de8bc1735accc494291e22da718bebab163547 ([`pr/libexec.7`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.7) -> [`pr/libexec.8`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.7..pr/libexec.8)) updating commit message to match release description, and implementing Sjors `bitcoin` help suggestion https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-2922787970 and release note suggestion https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2182003768
Updated 05de8bc1735accc494291e22da718bebab163547 -> 03f585d58ccf4d5c02d621c5b6046d45807b3201 ([`pr/libexec.8`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.8) -> [`pr/libexec.9`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.8..pr/libexec.9)) fixing mistake in release notes

Sjors reviewed (approved):
re-ACK 03f585d58ccf4d5c02d621c5b6046d45807b3201

Left a suggestion for slightly improved wording.

l0rinc commented on doc/files.md:
```diff
@@ -123,8 +123,40 @@ Subdirectory | File(s)           | Description
+
+## Installed Files
+
+This table describes the files installed by Bitcoin across different platforms.
```
```suggestion
This table describes the files installed by Bitcoin Core across different platforms.
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2185068952

Thanks! Updated

l0rinc commented on doc/files.md:
```diff
@@ -123,8 +123,40 @@ Subdirectory | File(s)           | Description
+### Notes
+
+- *Italicized* files are installed in source builds but not included in binary releases.
+- README and bitcoin.conf files included in binary releases but not installed in source builds.
```
nit: there's a similar one in the PR description:
> The table below shows the install location and availability **of** each binary after this change

l0rinc commented on doc/files.md:
```diff
@@ -16,7 +16,7 @@
   - [Berkeley DB database based wallets](#berkeley-db-database-based-wallets)
 
-- [Notes](#notes)
+- [Installed Files](#installed-files)
```
nit: if you're already editing this file consider fixing line 86:
> ... double-spends due to *the* lack of synchronization

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2185083314

> nit: if you're already editing this file consider fixing line 86:
> 
> > ... double-spends due to _the_ lack of synchronization

Thanks, added commit to fix

l0rinc commented on doc/release-notes-31375-31679.md:
```diff
@@ -0,0 +1,24 @@
+
+Other executables which are only part of source releases and not built by
+default: `test_bitcoin-qt`, `bench_bitcoin`, `bitcoin-chainstate`,
+`bitcoin-node`, and `bitcoin-wallet`  are also now installed in `libexec/`
```
`bitcoin-wallet` should still be in `bin` as far as I can tell, not `libexec` (nit: + double space):
```suggestion
`bitcoin-node` are also now installed in `libexec/`
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2185089423

> `bitcoin-wallet` should still be in `bin` as far as I can tell, not `libexec` (nit: + double space):

Good catch! I did mean to write `bitcoin-gui` not `bitcoin-wallet`

l0rinc commented on src/bitcoin.cpp:
```diff
@@ -63,11 +66,17 @@ int main(int argc, char* argv[])
             return EXIT_SUCCESS;
         }
 
+        std::string exe_name{fs::PathToString(fs::PathFromString(argv[0]).filename())};
```
Tested on Linux:
```bash
# cmake -B build -G Ninja && ninja -C build install && install/bin/bitcoin --help | head -1
...
Usage: bitcoin [OPTIONS] COMMAND...
```

l0rinc reviewed (approved):
Concept ACK, tested on Mac and Linux, it works as expected.

```bash
rm -rfd build install && \
cmake -B build -G Ninja \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DCMAKE_INSTALL_PREFIX="$PWD/install" \
  -DBUILD_TESTS=ON \
  -DBUILD_BENCH=ON \
  -DBUILD_FUZZ_BINARY=ON \
  -DBUILD_UTIL_CHAINSTATE=ON \
  -DBUILD_GUI=ON \
  -DBUILD_GUI_TESTS=ON && \
ninja -C build install && \
tree -L 3 install/{bin,libexec}
```

<details>
<summary>Mac</summary>

> -- Installing: bitcoin/install/bin/bitcoin-wallet
-- Installing: bitcoin/install/share/man/man1/bitcoin-wallet.1
-- Installing: bitcoin/install/bin/bitcoin
-- Installing: bitcoin/install/bin/bitcoind
-- Installing: bitcoin/install/share/man/man1/bitcoind.1
-- Installing: bitcoin/install/bin/bitcoin-cli
-- Installing: bitcoin/install/share/man/man1/bitcoin-cli.1
-- Installing: bitcoin/install/bin/bitcoin-tx
-- Installing: bitcoin/install/share/man/man1/bitcoin-tx.1
-- Installing: bitcoin/install/bin/bitcoin-util
-- Installing: bitcoin/install/share/man/man1/bitcoin-util.1
-- Installing: bitcoin/install/bin/bitcoin-qt
-- Installing: bitcoin/install/share/man/man1/bitcoin-qt.1
-- Installing: bitcoin/install/libexec/test_bitcoin-qt
-- Installing: bitcoin/install/lib/pkgconfig/libbitcoinkernel.pc
-- Installing: bitcoin/install/lib/libbitcoinkernel.dylib
-- Installing: bitcoin/install/libexec/bitcoin-chainstate
-- Installing: bitcoin/install/libexec/bench_bitcoin
-- Installing: bitcoin/install/libexec/test_bitcoin

> install/bin
├── bitcoin
├── bitcoin-cli
├── bitcoin-qt
├── bitcoin-tx
├── bitcoin-util
├── bitcoin-wallet
└── bitcoind
install/libexec
├── bench_bitcoin
├── bitcoin-chainstate
├── test_bitcoin
└── test_bitcoin-qt

</details>

<details>
<summary>Linux</summary>

> -- Installing: bitcoin/install/bin/bitcoin-wallet
-- Installing: bitcoin/install/share/man/man1/bitcoin-wallet.1
-- Installing: bitcoin/install/bin/bitcoin
-- Installing: bitcoin/install/bin/bitcoind
-- Installing: bitcoin/install/share/man/man1/bitcoind.1
-- Installing: bitcoin/install/bin/bitcoin-cli
-- Installing: bitcoin/install/share/man/man1/bitcoin-cli.1
-- Installing: bitcoin/install/bin/bitcoin-tx
-- Installing: bitcoin/install/share/man/man1/bitcoin-tx.1
-- Installing: bitcoin/install/bin/bitcoin-util
-- Installing: bitcoin/install/share/man/man1/bitcoin-util.1
-- Installing: bitcoin/install/bin/bitcoin-qt
-- Installing: bitcoin/install/share/man/man1/bitcoin-qt.1
-- Installing: bitcoin/install/libexec/test_bitcoin-qt
-- Installing: bitcoin/install/lib/pkgconfig/libbitcoinkernel.pc
-- Installing: bitcoin/install/lib/libbitcoinkernel.so
-- Installing: bitcoin/install/libexec/bitcoin-chainstate
-- Installing: bitcoin/install/libexec/bench_bitcoin
-- Installing: bitcoin/install/libexec/test_bitcoin

> install/bin
├── bitcoin
├── bitcoin-cli
├── bitcoind
├── bitcoin-qt
├── bitcoin-tx
├── bitcoin-util
└── bitcoin-wallet
install/libexec
├── bench_bitcoin
├── bitcoin-chainstate
├── test_bitcoin
└── test_bitcoin-qt

</details>

(only the `bitcoin` order and `libbitcoinkernel.dylib`vs`libbitcoinkernel.so` differ)

ryanofsky committed:
doc: Fix typo in files.md

Suggested by l0rinc https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2185083314

ryanofsky commented on ci/test/03_test_script.sh:
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2184948130

> Q: slightly unrelated, but if testing should be available for the installed version, how come fuzzing isn't? Seems it should be mentioned with a 🛠 at least...

The table doesn't mention the fuzz binary just because the table is a list of installed files, and the fuzz binary is not installed before or after this change.

Installing the fuzz binary would seem to make sense for consistency. But one problem with installing it as-is is that unlike `test_bitcoin`, `bench_bitcoin`, etc it has a generic name `fuzz` so I felt it would be better to leave alone here.

ryanofsky commented on ci/test/03_test_script.sh:
```diff
@@ -145,7 +145,7 @@ if [ "$RUN_UNIT_TESTS" = "true" ]; then
 
 if [ "$RUN_UNIT_TESTS_SEQUENTIAL" = "true" ]; then
-  DIR_UNIT_TEST_DATA="${DIR_UNIT_TEST_DATA}" LD_LIBRARY_PATH="${DEPENDS_DIR}/${HOST}/lib" "${BASE_OUTDIR}"/bin/test_bitcoin --catch_system_errors=no -l test_suite
+  DIR_UNIT_TEST_DATA="${DIR_UNIT_TEST_DATA}" LD_LIBRARY_PATH="${DEPENDS_DIR}/${HOST}/lib" "${BASE_OUTDIR}"/libexec/test_bitcoin --catch_system_errors=no -l test_suite
```
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2184924462

Thanks! I implemented maflcko's suggestion:

```diff
-  DIR_UNIT_TEST_DATA="${DIR_UNIT_TEST_DATA}" LD_LIBRARY_PATH="${DEPENDS_DIR}/${HOST}/lib" "${BASE_OUTDIR}"/libexec/test_bitcoin --catch_system_errors=no -l test_suite
+  DIR_UNIT_TEST_DATA="${DIR_UNIT_TEST_DATA}" LD_LIBRARY_PATH="${DEPENDS_DIR}/${HOST}/lib" "${BASE_BUILD_DIR}"/bin/test_bitcoin --catch_system_errors=no -l test_suite
```

But I am wondering if maybe it is actually good to run `test_bitcoin` from the install directory instead of the build directory, since maybe this could detect problems if unit test didn't work when installed? But happy to do anything here.

On the layout question, I agree it's nice for the build layout to mirror the install layout (even though this is unusual—usually build layout mirrors the source layout) and I implemented https://github.com/ryanofsky/bitcoin/commit/6e3bfb2b6df4774c1009d9210bc38f6db1a83c6d as a followup to support that.

I do think this PR should be focused on installation and packaging and not make changes that affect developers because concerns are different in the build and install cases.

For example, the install `bin/` directory usually contains binaries from dozens of packages, while `build/bin/` only contains bitcoin binaries. And the install `bin/` directory rarely needs to be referenced directly since it is in the `PATH`, while `build/bin/` is something developers will frequently type and using in scripts.

So I want to address packaging only here and avoid changes that could impact developers. But I am interested in https://github.com/ryanofsky/bitcoin/commit/6e3bfb2b6df4774c1009d9210bc38f6db1a83c6d as a followup and happy to discuss it  since I would like to keep build and install layouts consistent if it makes sense, even though they are controlled separately and used differently.

ryanofsky commented on doc/files.md:
```diff
@@ -123,8 +123,40 @@ Subdirectory | File(s)           | Description
+| libexec/test_bitcoin                | Unit test binary |
+| *libexec/test_bitcoin-qt*           | GUI-specific unit tests |
+| share/man/man1/                     | Man pages for command-line tools like `bitcoin-cli`, `bitcoind`, and others |
+| share/rpcauth/                      | Documentation and scripts for RPC authentication setup |
```
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2184944417

> nit: not everyone views the rendered view, we might want to align the source view as well:

Thanks! Updated

Sjors replied:
Looks great with `mdcat` :-)

ryanofsky commented on doc/release-notes-31375-31679.md:
```diff
@@ -0,0 +1,24 @@
+Install changes
+---------------
+
+The `test_bitcoin` executable is now located in `libexec/` rather than `bin/`.
+It can still be executed directly, or accessed through the new `bitcoin` command
+line tool as `bitcoin test`.
```
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2184938620

Good catch! Updated these as well

ryanofsky commented on src/bitcoin.cpp:
```diff
@@ -31,22 +31,25 @@ static constexpr auto HELP_USAGE = R"(Usage: %s [OPTIONS] COMMAND...
 )";
 
-static constexpr auto HELP_EXTRA = R"(
+static constexpr auto HELP_EXTRA = R"(  help           Show this help message.
```
re: https://github.com/bitcoin/bitcoin/pull/31679#discussion_r2184561400

> That's because if a user starts with "bitcoin" then "this help message" implies there's nothing additional to see.

To be sure, the "show this help message" does not appear when you just run `bitcoin` with no arguments since I wanted "show this help message" to mean "show this exact help message". Happy to apply any changes, but I think this should already avoid the concern you described.

ryanofsky reviewed (commented):
Updated 03f585d58ccf4d5c02d621c5b6046d45807b3201 -> 6a6b6f9093c1f82b8ea05ad87f421eabc8849738 ([`pr/libexec.9`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.9) -> [`pr/libexec.10`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.9..pr/libexec.10)) with various suggestions updating signing & CI scripts and documentation

ryanofsky reviewed (commented):
Updated 6a6b6f9093c1f82b8ea05ad87f421eabc8849738 -> f49840dd902cd9b14b6aadb431b16a4aeb719c3f ([`pr/libexec.10`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.10) -> [`pr/libexec.11`](https://github.com/ryanofsky/bitcoin/commits/pr/libexec.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/libexec.10..pr/libexec.11)) addressing remaining comments and making `bitcoin` help output more consistent

janb84 reviewed (commented):
re ACK f49840dd902cd9b14b6aadb431b16a4aeb719c3f

Changes since last ACK: 

- changes related to various suggestions
- help output improvements:
<details>

```shell
$  ./build/bin/bitcoin
Usage: bitcoin [OPTIONS] COMMAND...

Options:
  -m, --multiprocess     Run multiprocess binaries bitcoin-node, bitcoin-gui.
  -M, --monolithic       Run monolithic binaries bitcoind, bitcoin-qt. (Default behavior)
  -v, --version          Show version information
  -h, --help             Show full help message

Commands:
  gui [ARGS]     Start GUI, equivalent to running 'bitcoin-qt [ARGS]' or 'bitcoin-gui [ARGS]'.
  node [ARGS]    Start node, equivalent to running 'bitcoind [ARGS]' or 'bitcoin-node [ARGS]'.
  rpc [ARGS]     Call RPC method, equivalent to running 'bitcoin-cli -named [ARGS]'.
  wallet [ARGS]  Call wallet command, equivalent to running 'bitcoin-wallet [ARGS]'.
  tx [ARGS]      Manipulate hex-encoded transactions, equivalent to running 'bitcoin-tx [ARGS]'.
  help           Show full help message.

Run 'bitcoin help' to see additional commands (e.g. for testing and debugging).
```

```shell
$  ./build/bin/bitcoin help
Usage: bitcoin [OPTIONS] COMMAND...

Options:
  -m, --multiprocess     Run multiprocess binaries bitcoin-node, bitcoin-gui.
  -M, --monolithic       Run monolithic binaries bitcoind, bitcoin-qt. (Default behavior)
  -v, --version          Show version information
  -h, --help             Show full help message

Commands:
  gui [ARGS]     Start GUI, equivalent to running 'bitcoin-qt [ARGS]' or 'bitcoin-gui [ARGS]'.
  node [ARGS]    Start node, equivalent to running 'bitcoind [ARGS]' or 'bitcoin-node [ARGS]'.
  rpc [ARGS]     Call RPC method, equivalent to running 'bitcoin-cli -named [ARGS]'.
  wallet [ARGS]  Call wallet command, equivalent to running 'bitcoin-wallet [ARGS]'.
  tx [ARGS]      Manipulate hex-encoded transactions, equivalent to running 'bitcoin-tx [ARGS]'.
  help           Show full help message.

Additional less commonly used commands:
  bench [ARGS]      Run bench command, equivalent to running 'bench_bitcoin [ARGS]'.
  chainstate [ARGS] Run bitcoin kernel chainstate util, equivalent to running 'bitcoin-chainstate [ARGS]'.
  test [ARGS]       Run unit tests, equivalent to running 'test_bitcoin [ARGS]'.
  test-gui [ARGS]   Run GUI unit tests, equivalent to running 'test_bitcoin-qt [ARGS]'.
```
</details>

l0rinc commented:
re-ACK f49840dd902cd9b14b6aadb431b16a4aeb719c3f

BrandonOdiwuor reviewed (commented):
Tested ACK f49840dd902cd9b14b6aadb431b16a4aeb719c3f

Tested on macOS 15.5 that `bitcoin-node` and `test_bitcoin` have been correctly moved from `bin/` to `libexec/`

<img width="635" height="311" alt="Screenshot 2025-07-10 at 18 56 31" src="https://github.com/user-attachments/assets/7f0e6ae5-f06f-488f-b913-1494bd20c05a" />

---

**Update:**
Also tested that `bench_bitcoin`, `bitcoin_gui` and `test_bitcoin-qt` are also moved to `libexec/`
<img width="898" height="106" alt="Screenshot 2025-07-10 at 21 32 44" src="https://github.com/user-attachments/assets/4bc71233-c851-42a1-984b-86b27bde3d4b" />

hodlinator reviewed (approved):
re-ACK f49840dd902cd9b14b6aadb431b16a4aeb719c3f

Only able to spot improvements since previous ACK (https://github.com/bitcoin/bitcoin/pull/31679#pullrequestreview-2980338061). :)
Re-tested `cmake --install` + running `bitcoin` wrapper and `test_bitcoin`.

Agree with Sjors that codesigning should probably be re-tested since it's been modified (https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-3047748032).

Sjors commented:
@TheCharlatan better?

TheCharlatan commented:
I'm still not fully convinced of the new layout, but also would not unequivocally block this. Maybe some more long term contributors should give their approval here though, since it does change how our binaries are shipped and where we are placing things going forward.

Re Russ' previous comment:
> based on what a good command line interface would look like, not based on idiosyncracies of previous releases.

There is a bunch of nuance here, since this also applies to pretty much all the binaries remaning in /bin besides the wrapper executable. What the comment does seem to imply to me is that going forward all new binaries should be put into /libexec and be used through the wrapper binary to avoid idiosyncrasies.

TheCharlatan commented:
> I think it makes more sense that some of our binaries should be command line binaries presenting stable, backward-compatible CLIs, and other binaries can be internal binaries that do useful things for us but users are not expected to use directly and we are free to change, rename, merge, or split up in the future.

It's not clear to me how the approach here really helps with that though. Essentially it just adds a tiny extra step for the user to use these binaries. I think the utility of the change is more one of drawing a border for ourselves between stuff that can be expected to be stable and stuff that might change and break existing behaviour in the future. But I don't think this boundary is clearly communicated to the user by this change.

achow101 commented:
I'm not really a fan of changing binary locations yet again. If this is merged for v30.0, we'll have the 3 active major release branches all with different binary locations. I expect this will make my workflow a bit harder and make backports more annoying to test.

Mainly my complaint is about the locations of `bench_bitcoin` and `test_bitcoin` in builds from source since those end up being used quite a lot in my development workflow.

For the other (new) binaries that would be in libexec, I have no strong opinion, but at least those binaries aren't in the binary releases, and I doubt people are actively using multiprocess heavily in their development. However, if we could pick one location and stick with it for the rest of time, that'd be great.

***

Tangentially, is there a reason we ship `test_bitcoin` in the binary releases?

Sjors commented:
@achow101 there was some discussion about `test_bitcoin` above: https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-2807539249
https://github.com/bitcoin/bitcoin/pull/31679#issuecomment-3027233699

ryanofsky commented:
> This only affects the binary release and not source built locations, so actually, I don't care either way.

Thanks, yes. Effects of this PR:

* For **source builds** (i.e. developer builds) — No changes.
* For **source installs** (i.e. `cmake --install` output) — `test_bitcoin` and `test_bitcoin-qt` and `bench_bitcoin` are installed in `${CMAKE_PREFIX_PATH}/libexec` instead of `${CMAKE_PREFIX_PATH}/bin`, so they are no longer on `PATH`. But they can still be invoked from the `libexec/` directory or from the CLI as `bitcoin test`, `bitcoin test-gui`, `bitcoin bench`.
* For **binary releases** — Since `test_bitcoin` is the only test binary enabled in releases, the only change is moving `test_bitcoin` from `bin/` to `libexec/`.

The goal of the PR is to introduce a distinction between internal and external binaries so that starting with #31802, we can use IPC to provide features implemented in new binaries and new binaries do not automatically become part of the CLI. The same distinction is also useful for tests. We should be able to change the way tests are built without affecting the CLI.

willcl-ark reviewed (commented):
Concept ACK.

~I still see a number of references to out-dated locations in our codebase with this PR:~

```bash
$ git grep "bin/\(test_bitcoin\|bench_bitcoin\|bitcoin-chainstate\|bitcoin-gui\|bitcoin-node\|test_bitcoin-qt\)"
.github/workflows/ci.yml:          ./bin/test_bitcoin.exe -l test_suite
.github/workflows/ci.yml:        run: ./bin/bench_bitcoin.exe -sanity-check
ci/test/03_test_script.sh:  DIR_UNIT_TEST_DATA="${DIR_UNIT_TEST_DATA}" LD_LIBRARY_PATH="${DEPENDS_DIR}/${HOST}/lib" "${BASE_BUILD_DIR}"/bin/test_bitcoin --catch_system_errors=no -l test_suite
contrib/devtools/deterministic-unittest-coverage/src/main.rs:    let test_exe = build_dir.join("bin/test_bitcoin");
contrib/valgrind.supp:# $ valgrind --suppressions=contrib/valgrind.supp build/bin/test_bitcoin
contrib/valgrind.supp:#       --show-leak-kinds=all build/bin/test_bitcoin
contrib/valgrind.supp:#       --error-limit=no build/bin/test_bitcoin
doc/benchmarking.md:    build/bin/bench_bitcoin
doc/benchmarking.md:    build/bin/bench_bitcoin -h
doc/developer-notes.md:$ valgrind --suppressions=contrib/valgrind.supp build/bin/test_bitcoin
doc/developer-notes.md:      --show-leak-kinds=all build/bin/test_bitcoin --log_level=test_suite
doc/developer-notes.md:    --object=build/bin/test_bitcoin \
src/qt/test/wallettests.cpp://     QT_QPA_PLATFORM=xcb     build/bin/test_bitcoin-qt  # Linux
src/qt/test/wallettests.cpp://     QT_QPA_PLATFORM=windows build/bin/test_bitcoin-qt  # Windows
src/qt/test/wallettests.cpp://     QT_QPA_PLATFORM=cocoa   build/bin/test_bitcoin-qt  # macOS
src/test/README.md:Run `build/bin/test_bitcoin --list_content` for the full list of tests.
src/test/README.md:To run the unit tests manually, launch `build/bin/test_bitcoin`. To recompile
src/test/README.md:To run the GUI unit tests manually, launch `build/bin/test_bitcoin-qt`
src/test/README.md:build/bin/test_bitcoin --help
src/test/README.md:build/bin/test_bitcoin --log_level=all --run_test=getarg_tests
src/test/README.md:build/bin/test_bitcoin -l all -t getarg_tests
src/test/README.md:build/bin/test_bitcoin --run_test=getarg_tests/doubledash
src/test/README.md:build/bin/test_bitcoin --log_level=all --run_test=getarg_tests -- -printtoconsole=1
src/test/README.md:$ build/bin/test_bitcoin --run_test=getarg_tests/doubledash -- -testdatadir=/somewhere/mydatadir
src/test/README.md:gdb build/bin/test_bitcoin
src/test/README.md:is happening by running `gdb ./build/bin/test_bitcoin` and then using the `bt` command
src/test/README.md:gdb build/bin/test_bitcoin core
```

~Not all of these need fixing, but it would seem nice to get the ones that should be fixed done here, or at least noted for a followup (if you want to try and avoid invalidating 5 ACKs).~

Actually looking closer these are pretty much all build directory links, so are ok.

ryanofsky commented:
> Actually looking closer these are pretty much all build directory links, so are ok.

Sorry about that, the PR description was confusing and achow thought the same thing you did. I rewrote it now to be clearer. (I think the original PR description was also clear but I had replaced it a few weeks ago with a big table to explain some details of the change, and I think the table made it harder to understand the high-level change.)

willcl-ark reviewed (approved):
utACK f49840dd902cd9b14b6aadb431b16a4aeb719c3f

> Sorry about that, the PR description was confusing ...

No that was my bad, I should have read the PR comments before reviewing (perhaps, although I don't like to).

I am in favour of moving binaries to libexec/, including `test_bitcoin`, as per this change.

I've tried to find other bitcoin packages which may break on this, but can't find many, and the few I have found are already outdated in other ways, so require a "manual update" already.

I wondered if this might break any of our shell completions, but none of those are affected by this PR which is great.  (I did notice in checking that we don't have completions for the  `bitcoin` program.)

<details>
<summary>Guix build</summary>

```
❯ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
9080e91998fa6f384c00a2e3aab6316fdd2174a89a303b0a602880daf515defb  guix-build-f49840dd902c/output/aarch64-linux-gnu/SHA256SUMS.part
c4061ae7a239ea148d2423f7414d9dd03cefbdfd25c2142476fc400b57ccfef8  guix-build-f49840dd902c/output/aarch64-linux-gnu/bitcoin-f49840dd902c-aarch64-linux-gnu-debug.tar.gz
878588692759d57a0ebb00f3ceded73c53e0c2d9f53dbd0a83549a22d48c007c  guix-build-f49840dd902c/output/aarch64-linux-gnu/bitcoin-f49840dd902c-aarch64-linux-gnu.tar.gz
1ce7c5b093514d2d9caf67477dded186bbc1fc740318d7b14c9e7a98d27b5e20  guix-build-f49840dd902c/output/arm-linux-gnueabihf/SHA256SUMS.part
d796ebc0b64cb6b16bbb729dc444106464748069124811ac1430e48037045585  guix-build-f49840dd902c/output/arm-linux-gnueabihf/bitcoin-f49840dd902c-arm-linux-gnueabihf-debug.tar.gz
6e911e00bd918e3837bc129ecd8e8a6071d7930cbeb4c4ea3a5d602b156d17a7  guix-build-f49840dd902c/output/arm-linux-gnueabihf/bitcoin-f49840dd902c-arm-linux-gnueabihf.tar.gz
6d40e2758b092acb5ec1745ef2f85c666c61a8f04fb89c996ec1f452db66fc3f  guix-build-f49840dd902c/output/arm64-apple-darwin/SHA256SUMS.part
7ca976099f3d02d68b3806c3a51203a8968e8f1f3ce3cba43cca2fd7c0f24c29  guix-build-f49840dd902c/output/arm64-apple-darwin/bitcoin-f49840dd902c-arm64-apple-darwin-codesigning.tar.gz
3c9de7f8c734b47809011a4a64668c4c19b7452cf0f963b9861769d9e4729909  guix-build-f49840dd902c/output/arm64-apple-darwin/bitcoin-f49840dd902c-arm64-apple-darwin-unsigned.tar.gz
f3c313c5847ddf763cf751b859c40145be5f5c074b46491a4e4665cff577af33  guix-build-f49840dd902c/output/arm64-apple-darwin/bitcoin-f49840dd902c-arm64-apple-darwin-unsigned.zip
17ae536221bccec80d53e9c4b5512095be6f97c44957ba836841f450b6c44dbc  guix-build-f49840dd902c/output/dist-archive/bitcoin-f49840dd902c.tar.gz
8025a760fd94c9a8473aba4b769f28088224b4a7594e543c59ba474e48626df5  guix-build-f49840dd902c/output/powerpc64-linux-gnu/SHA256SUMS.part
79b05779161c43c0ef4177e09e6b200510efe02beeb0d4e45ce623ffc12750ec  guix-build-f49840dd902c/output/powerpc64-linux-gnu/bitcoin-f49840dd902c-powerpc64-linux-gnu-debug.tar.gz
59d5cdd24475611e77d7402e4770ef0c9bfb630827d1adb70a26eb0b42f15503  guix-build-f49840dd902c/output/powerpc64-linux-gnu/bitcoin-f49840dd902c-powerpc64-linux-gnu.tar.gz
853fd79a50cc44202d17e7cc3d3eff33d06c28aafe2a652060fa6ad25d8d9446  guix-build-f49840dd902c/output/riscv64-linux-gnu/SHA256SUMS.part
acb6f097a16f06bdeb0687d859c0db77ceb8c2d39e1981b0777a332034ce5f63  guix-build-f49840dd902c/output/riscv64-linux-gnu/bitcoin-f49840dd902c-riscv64-linux-gnu-debug.tar.gz
815f7870ca273dd5578b592f1815043abcd05f92ff8f3310eeb5abea8bd1dbf4  guix-build-f49840dd902c/output/riscv64-linux-gnu/bitcoin-f49840dd902c-riscv64-linux-gnu.tar.gz
85117256abc78980e928d45453453805807b189e3fb3555d1f834e8d8aa2957a  guix-build-f49840dd902c/output/x86_64-apple-darwin/SHA256SUMS.part
de9493f16cceb93bb59b0af259778bef9a1f677651a60736eb1a445c2c10380b  guix-build-f49840dd902c/output/x86_64-apple-darwin/bitcoin-f49840dd902c-x86_64-apple-darwin-codesigning.tar.gz
6febb9c4420c720361c9e590499564e28550da8f64c61bff3a68f35890195939  guix-build-f49840dd902c/output/x86_64-apple-darwin/bitcoin-f49840dd902c-x86_64-apple-darwin-unsigned.tar.gz
cf948b94e2a6dab9dbbc5f2a2108a0cb7f2639cf3965c7c390cdb3918997b252  guix-build-f49840dd902c/output/x86_64-apple-darwin/bitcoin-f49840dd902c-x86_64-apple-darwin-unsigned.zip
71b8a2b121a3e83854649b0e07df161b3cc49bef95c17fb901c1171a7b42db3d  guix-build-f49840dd902c/output/x86_64-linux-gnu/SHA256SUMS.part
2469fda346eb8e1979e13e0deb1828861c50b64e747f2f004fe3c533c77ae581  guix-build-f49840dd902c/output/x86_64-linux-gnu/bitcoin-f49840dd902c-x86_64-linux-gnu-debug.tar.gz
e05dad8ad152c86dad619cfde1b62f341b8b855e6eca9784ec363d47e396d7d7  guix-build-f49840dd902c/output/x86_64-linux-gnu/bitcoin-f49840dd902c-x86_64-linux-gnu.tar.gz
07c3a2c8b9b7ef92d798c4c6020aa596faa3ac69f0726f2b39959d1ea411cae1  guix-build-f49840dd902c/output/x86_64-w64-mingw32/SHA256SUMS.part
77a2d8e7defce7f071d08a3aaa506c6d0aebf205cee244464552de8ff5c03bac  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-codesigning.tar.gz
9d2c8b6f9ab8ef31c90311ac058969cf98bcd3e67a414cc4cd4de2881e7bc3df  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-debug.zip
a60db9d3311e78c34a578a3fa7864dc34c37b5ca8cdc95371161c39d46f134ee  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-setup-unsigned.exe
76055f9d5b3f47d47cb798b8b5622543048d7f55f447ac7a2cb15da071caf2a1  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-unsigned.zip
```

</details>

achow101 commented:
ACK f49840dd902cd9b14b6aadb431b16a4aeb719c3f

```
$ find guix-build-f49840dd902c/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum                                             01:40:49 PM
9080e91998fa6f384c00a2e3aab6316fdd2174a89a303b0a602880daf515defb  guix-build-f49840dd902c/output/aarch64-linux-gnu/SHA256SUMS.part
c4061ae7a239ea148d2423f7414d9dd03cefbdfd25c2142476fc400b57ccfef8  guix-build-f49840dd902c/output/aarch64-linux-gnu/bitcoin-f49840dd902c-aarch64-linux-gnu-debug.tar.gz
878588692759d57a0ebb00f3ceded73c53e0c2d9f53dbd0a83549a22d48c007c  guix-build-f49840dd902c/output/aarch64-linux-gnu/bitcoin-f49840dd902c-aarch64-linux-gnu.tar.gz
1ce7c5b093514d2d9caf67477dded186bbc1fc740318d7b14c9e7a98d27b5e20  guix-build-f49840dd902c/output/arm-linux-gnueabihf/SHA256SUMS.part
d796ebc0b64cb6b16bbb729dc444106464748069124811ac1430e48037045585  guix-build-f49840dd902c/output/arm-linux-gnueabihf/bitcoin-f49840dd902c-arm-linux-gnueabihf-debug.tar.gz
6e911e00bd918e3837bc129ecd8e8a6071d7930cbeb4c4ea3a5d602b156d17a7  guix-build-f49840dd902c/output/arm-linux-gnueabihf/bitcoin-f49840dd902c-arm-linux-gnueabihf.tar.gz
6d40e2758b092acb5ec1745ef2f85c666c61a8f04fb89c996ec1f452db66fc3f  guix-build-f49840dd902c/output/arm64-apple-darwin/SHA256SUMS.part
7ca976099f3d02d68b3806c3a51203a8968e8f1f3ce3cba43cca2fd7c0f24c29  guix-build-f49840dd902c/output/arm64-apple-darwin/bitcoin-f49840dd902c-arm64-apple-darwin-codesigning.tar.gz
3c9de7f8c734b47809011a4a64668c4c19b7452cf0f963b9861769d9e4729909  guix-build-f49840dd902c/output/arm64-apple-darwin/bitcoin-f49840dd902c-arm64-apple-darwin-unsigned.tar.gz
f3c313c5847ddf763cf751b859c40145be5f5c074b46491a4e4665cff577af33  guix-build-f49840dd902c/output/arm64-apple-darwin/bitcoin-f49840dd902c-arm64-apple-darwin-unsigned.zip
17ae536221bccec80d53e9c4b5512095be6f97c44957ba836841f450b6c44dbc  guix-build-f49840dd902c/output/dist-archive/bitcoin-f49840dd902c.tar.gz
8025a760fd94c9a8473aba4b769f28088224b4a7594e543c59ba474e48626df5  guix-build-f49840dd902c/output/powerpc64-linux-gnu/SHA256SUMS.part
79b05779161c43c0ef4177e09e6b200510efe02beeb0d4e45ce623ffc12750ec  guix-build-f49840dd902c/output/powerpc64-linux-gnu/bitcoin-f49840dd902c-powerpc64-linux-gnu-debug.tar.gz
59d5cdd24475611e77d7402e4770ef0c9bfb630827d1adb70a26eb0b42f15503  guix-build-f49840dd902c/output/powerpc64-linux-gnu/bitcoin-f49840dd902c-powerpc64-linux-gnu.tar.gz
853fd79a50cc44202d17e7cc3d3eff33d06c28aafe2a652060fa6ad25d8d9446  guix-build-f49840dd902c/output/riscv64-linux-gnu/SHA256SUMS.part
acb6f097a16f06bdeb0687d859c0db77ceb8c2d39e1981b0777a332034ce5f63  guix-build-f49840dd902c/output/riscv64-linux-gnu/bitcoin-f49840dd902c-riscv64-linux-gnu-debug.tar.gz
815f7870ca273dd5578b592f1815043abcd05f92ff8f3310eeb5abea8bd1dbf4  guix-build-f49840dd902c/output/riscv64-linux-gnu/bitcoin-f49840dd902c-riscv64-linux-gnu.tar.gz
85117256abc78980e928d45453453805807b189e3fb3555d1f834e8d8aa2957a  guix-build-f49840dd902c/output/x86_64-apple-darwin/SHA256SUMS.part
de9493f16cceb93bb59b0af259778bef9a1f677651a60736eb1a445c2c10380b  guix-build-f49840dd902c/output/x86_64-apple-darwin/bitcoin-f49840dd902c-x86_64-apple-darwin-codesigning.tar.gz
6febb9c4420c720361c9e590499564e28550da8f64c61bff3a68f35890195939  guix-build-f49840dd902c/output/x86_64-apple-darwin/bitcoin-f49840dd902c-x86_64-apple-darwin-unsigned.tar.gz
cf948b94e2a6dab9dbbc5f2a2108a0cb7f2639cf3965c7c390cdb3918997b252  guix-build-f49840dd902c/output/x86_64-apple-darwin/bitcoin-f49840dd902c-x86_64-apple-darwin-unsigned.zip
71b8a2b121a3e83854649b0e07df161b3cc49bef95c17fb901c1171a7b42db3d  guix-build-f49840dd902c/output/x86_64-linux-gnu/SHA256SUMS.part
2469fda346eb8e1979e13e0deb1828861c50b64e747f2f004fe3c533c77ae581  guix-build-f49840dd902c/output/x86_64-linux-gnu/bitcoin-f49840dd902c-x86_64-linux-gnu-debug.tar.gz
e05dad8ad152c86dad619cfde1b62f341b8b855e6eca9784ec363d47e396d7d7  guix-build-f49840dd902c/output/x86_64-linux-gnu/bitcoin-f49840dd902c-x86_64-linux-gnu.tar.gz
07c3a2c8b9b7ef92d798c4c6020aa596faa3ac69f0726f2b39959d1ea411cae1  guix-build-f49840dd902c/output/x86_64-w64-mingw32/SHA256SUMS.part
77a2d8e7defce7f071d08a3aaa506c6d0aebf205cee244464552de8ff5c03bac  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-codesigning.tar.gz
9d2c8b6f9ab8ef31c90311ac058969cf98bcd3e67a414cc4cd4de2881e7bc3df  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-debug.zip
a60db9d3311e78c34a578a3fa7864dc34c37b5ca8cdc95371161c39d46f134ee  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-setup-unsigned.exe
76055f9d5b3f47d47cb798b8b5622543048d7f55f447ac7a2cb15da071caf2a1  guix-build-f49840dd902c/output/x86_64-w64-mingw32/bitcoin-f49840dd902c-win64-unsigned.zip
```
